src/ekore/harmonics/cache.py

```diff
@@ -0,0 +1,229 @@
+"""Caching harmonic sums across :mod:`ekore`."""
+from typing import Optional
+
+import numba as nb
+import numpy as np
+import numpy.typing as npt
+
+from . import w1, w2, w3, w4, w5
+from .g_functions import mellin_g3
+from .polygamma import recursive_harmonic_sum
+
+# here a register of all possible functions
+CACHE_SIZE = 31
+_index = iter(range(CACHE_SIZE))
+S1 = next(_index)  # = S_1(N)
+S2 = next(_index)
+S3 = next(_index)
+S4 = next(_index)
+S5 = next(_index)
+Sm1 = next(_index)
+Sm2 = next(_index)
+Sm3 = next(_index)
+Sm4 = next(_index)
+Sm5 = next(_index)
+S21 = next(_index)
+S2m1 = next(_index)
+Sm21 = next(_index)
+Sm2m1 = next(_index)
+S31 = next(_index)
+Sm31 = next(_index)
+Sm22 = next(_index)
+S211 = next(_index)
+Sm211 = next(_index)
+S1h = next(_index)
+S2h = next(_index)
+S3h = next(_index)
+S1mh = next(_index)
+S2mh = next(_index)
+S3mh = next(_index)
+S1ph = next(_index)
+S2ph = next(_index)
+S3ph = next(_index)
+g3 = next(_index)
+S1p2 = next(_index)
+g3p2 = next(_index)
+
+
+@nb.njit(cache=True)
+def reset():
+    """Return the cache placeholder array."""
+    return np.full(CACHE_SIZE, np.nan, np.complex_)
+
+
+@nb.njit(cache=True)
+def update(func, key, cache, n):
+    """Compute simple harmonics if not yet in cache."""
+    if np.isnan(cache[key]):
+        cache[key] = func(n)
+    return cache
+
+
+@nb.njit(cache=True)
+def update_Sm1(cache, n, is_singlet):
+    """Compute Sm1 if not yet in cache."""
+    if np.isnan(cache[Sm1]):
+        cache = update(w1.S1, S1, cache, n)
+        cache = update(w1.S1, S1mh, cache, (n - 1) / 2)
+        cache = update(w1.S1, S1h, cache, n / 2)
+        cache[Sm1] = w1.Sm1(n, cache[S1], cache[S1mh], cache[S1h], is_singlet)
+    return cache
+
+
+@nb.njit(cache=True)
+def update_Sm2(cache, n, is_singlet):
+    """Compute Sm2 if not yet in cache."""
+    if np.isnan(cache[Sm2]):
+        cache = update(w2.S2, S2, cache, n)
+        cache = update(w2.S2, S2mh, cache, (n - 1) / 2)
+        cache = update(w2.S2, S2h, cache, n / 2)
+        cache[Sm2] = w2.Sm2(n, cache[S2], cache[S2mh], cache[S2h], is_singlet)
+    return cache
+
+
+@nb.njit(cache=True)
```

Comment:
```suggestion
@nb.njit
```
https://github.com/NNPDF/eko/pull/253#discussion_r1183392401

Reply:
It is true that everything that is used under Numba has to be numbified, but in principle this property doesn't extend to cache: we should be able to mix cached and uncached Numba functions.

Since the problem seems to derive from this one, let's simply move this single function out of cache.

Reply:
The following seems to work fine:
```py
import numba as nb

@nb.njit(cache=True)
def callee(x):
    return x == 0

@nb.njit
def caller(exit):
    if exit:
        return 1.
    return callee(caller(0))

@nb.njit(cache=True)
def supercaller():
    return caller(True)

print(supercaller())
```

Reply:
Sorry, I take it back: if you replace `caller(True)` with `caller(False)` in `supercaller`, it is crashing again.

Reply:
yes, from what I saw yesterday as soon as you have a not cached functions all the `supercallers` cache will not work.